npm/webpack-dev-server/src/plugin.ts

```diff
@@ -52,85 +53,98 @@ export default class CypressCTOptionsPlugin {
     this.devServerEvents = options.devServerEvents
   }
 
-  private pluginFunc = (context: CypressCTWebpackContext) => {
-    context._cypress = {
+  private addLoaderContext = (loaderContext: object, module: any) => {
+    (loaderContext as CypressCTWebpackContext)._cypress = {
       files: this.files,
       projectRoot: this.projectRoot,
       supportFile: this.supportFile,
     }
   };
 
-  private setupCustomHMR = (compiler: webpack.Compiler) => {
-    compiler.hooks.afterCompile.tap(
-      'CypressCTOptionsPlugin',
-      (compilation) => {
-        const stats = compilation.getStats()
-
-        if (stats.hasErrors()) {
-          this.errorEmitted = true
-
-          // webpack 4: string[]
-          // webpack 5: Error[]
-          const errors = stats.toJson().errors as Array<Error | string> | undefined
-
-          if (!errors || !errors.length) {
-            return
-          }
-
-          this.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
-        } else if (this.errorEmitted) {
-          // compilation succeed but assets haven't emitted to the output yet
-          this.devServerEvents.emit('dev-server:compile:error', null)
-        }
-      },
-    )
-
-    compiler.hooks.afterEmit.tap(
-      'CypressCTOptionsPlugin',
-      (compilation) => {
-        if (!compilation.getStats().hasErrors()) {
-          this.devServerEvents.emit('dev-server:compile:success')
-        }
-      },
-    )
+  /*
+   * After compiling, we check for errors and inform the server of them.
+   */
+  private afterCompile = () => {
+    if (!this.compilation) {
+      return
+    }
+
+    const stats = this.compilation.getStats()
+
+    if (stats.hasErrors()) {
+      this.errorEmitted = true
+
+      // webpack 4: string[]
+      // webpack 5: Error[]
+      const errors = stats.toJson().errors as Array<Error | string> | undefined
+
+      if (!errors || !errors.length) {
+        return
+      }
+
+      this.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
+    } else if (this.errorEmitted) {
+      // compilation succeed but assets haven't emitted to the output yet
+      this.devServerEvents.emit('dev-server:compile:error', null)
+    }
+  }
+
+  // After emitting assets, we tell the server complitation was successful
+  // so it can trigger a reload the AUT iframe.
+  private afterEmit = () => {
+    if (!this.compilation?.getStats().hasErrors()) {
+      this.devServerEvents.emit('dev-server:compile:success')
+    }
+  }
+
+  /*
+   * `webpack --watch` watches the existing specs and their dependencies for changes,
+   * but we also need to add additional dependencies to our dynamic "browser.js" (generated
+   * using loader.ts) when new specs are created. This hook informs webpack that browser.js
+   * has been "updated on disk", causing a recompliation (and pulling the new specs in as
+   * dependencies).
+   */
+  private onSpecsChange = (specs: Cypress.Cypress['spec'][]) => {
```

Comment:
Nit: Can be more concise: `Cypress.Spec[]`

Reply:
I didn't think much of it, have just been copy-pasting this around from the existing types. I leave cleanup as an exercise to the reader. :D